src/os/bluestore/BlueStore_DB_Hash.cc

```diff
@@ -0,0 +1,678 @@
+      shards_it = db_hash.shards.begin();
+      if (shards_it == db_hash.shards.end())
+        return -1;
+      if (open_shards())
+        return 0;
+      else
+        return -1;
+    }
+
+    int seek_to_first(const string &prefix) override {
+      shards_it = db_hash.shards.find(prefix);
+      if (shards_it == db_hash.shards.end())
+        return -1;
+      if (open_shards())
+        return 0;
+      else
+        return -1;
+    }
+
+    int seek_to_last() override {
+      ceph_assert(false && "expected seek_to_last() not called");
+      return 0;
+    }
+
+    int seek_to_last(const string &prefix) override {
+      ceph_assert(false && "expected seek_to_last() not called");
+      return 0;
+    }
+
+    int upper_bound(const string &prefix, const string &after) override {
+      int r = seek_to_first(prefix);
+      if (r < 0)
+        return r;
+      for(auto& it: current_shards_iterators) {
+        it->upper_bound(after);
+      }
+      std::sort(current_shards_iterators.begin(), current_shards_iterators.end(), KeyLess(*this));
+      return 0;
+    }
+
+    int lower_bound(const string &prefix, const string &to) override {
+      int r = seek_to_first(prefix);
+      if (r < 0)
+        return r;
+      for(auto& it: current_shards_iterators) {
+        it->lower_bound(to);
+      }
+      struct KeyLess {
+        WholeSpaceIteratorMerged_Impl& iter;
+        KeyLess(WholeSpaceIteratorMerged_Impl& iter) : iter(iter) {};
+        bool operator()(KeyValueDB::Iterator a, KeyValueDB::Iterator b) const
+        {
+          if (!a->valid())
+            return false;
+          return iter.compare(a->key(), b->key()) < 0;
+        }
+      };
+      std::sort(current_shards_iterators.begin(), current_shards_iterators.end(), KeyLess(*this));
+      return 0;
+    }
+
+    bool valid() override {
+      if ((position < 0) || (position >= (ssize_t)current_shards_iterators.size()) )
+        return false;
+      return current_shards_iterators[position]->valid();
+    }
+
+    int next() override {
+      current_shards_iterators[position]->next();
+      if (current_shards_iterators[position]->valid()) {
+        /* this means that next element on this iterator is ok,
+         * but it is likely that it will NOT be next in order */
+        if (position == (ssize_t)current_shards_iterators.size() - 1) {
+          /* this is last one, so it must be correct one */
+          return 0;
+        }
+        std::string key0 = current_shards_iterators[position]->key();
+        for (size_t p = position + 1; p < current_shards_iterators.size(); p++) {
+          std::string key1 = current_shards_iterators[p]->key();
+          if (compare(key0, key1) < 0) {
+            /* all in order, no need to sort more */
+            break;
+          }
+          std::swap(current_shards_iterators[p - 1], current_shards_iterators[p]);
+        }
+        return 0;
+      }
+      ceph_assert(!current_shards_iterators[position]->valid());
```

Comment:
maybe just an ``else`` and comment, since this is exactly the condition on the ``if`` above?